File: python/janus/chat/room.py
"""ChatRoom - Main orchestrator for interactive multi-LLM chat."""

import asyncio
import re
from typing import TYPE_CHECKING

from rich.console import Console

from ..core.types import ChatConfig, ChatMessage, MessageRole
from ..providers.base import CompletionOptions, Provider
from .commands import CommandHandler
from .prompts import get_system_prompt, make_expand_prompt
from .session import ChatSession
from .ui import ChatDisplay


class ChatRoom:
    """Orchestrates interactive multi-LLM chat sessions."""

    def __init__(
        self,
        providers: list[Provider],
        config: ChatConfig | None = None,
        session: ChatSession | None = None,
        console: Console | None = None,
    ):
        self.providers = providers
        self.config = config or ChatConfig()
        self.session = session or ChatSession(
            active_models=[p.name for p in providers]
        )
        self.display = ChatDisplay(console)
        self.command_handler = CommandHandler(self)

    async def start(self) -> None:
        """Enter the main interactive loop."""
        self.display.show_welcome([p.name for p in self.providers])

        while True:
            try:
                # Get user input
                user_input = self.display.get_input().strip()

                if not user_input:
                    continue

                # Check for command
                command = self.command_handler.parse(user_input)
                if command:
                    result = await self.command_handler.execute(command)
                    if result.message:
                        self.display.show_command_result(result.message)
                    if result.should_exit:
                        break
                    continue

                # Regular message - send to models
                await self.send_message(user_input)

            except KeyboardInterrupt:
                self.display.console.print("\n[dim]Use /quit to exit[/dim]")
            except Exception as e:
                self.display.show_error(str(e))

    async def send_message(
        self,
        content: str,
        target_models: list[str] | None = None,
        expand: bool = False,
    ) -> None:
        """Send user message and collect model responses."""
        # Parse @mentions from content
        mentioned_models, clean_content = self._parse_mentions(content)

        # Determine which models should respond
        if target_models:
            responding = target_models
        elif mentioned_models:
            responding = mentioned_models
        else:
            responding = [p.name.lower() for p in self.providers]

        # Add user message to session (use original content with mentions)
        self.session.add_user_message(content)
        self.display.show_user_message(content)

        # Get responding providers
        responding_providers = [
            p for p in self.providers if p.name.lower() in responding
        ]

        if not responding_providers:
            self.display.show_error("No matching models found.")
            return

        # Show thinking indicator
        self.display.show_thinking([p.name for p in responding_providers])

        # Get responses
        if self.config.parallel_responses:
            tasks = [
                self._get_response(p, clean_content, expand)
                for p in responding_providers
            ]
            responses = await asyncio.gather(*tasks)
        else:
            responses = []
            for p in responding_providers:
                resp = await self._get_response(p, clean_content, expand)
                responses.append(resp)

        # Clear thinking indicator
        self.display.clear_thinking()

        # Display and store responses
        for response in responses:
            if response:
                self.session.add_message(response)
                self.display.show_model_response(response.model, response.content)

    async def _get_response(
        self,
        provider: Provider,
        content: str,
        expand: bool = False,
    ) -> ChatMessage | None:
        """Get a single model's response."""
        try:
            # Build context from conversation history
            context = self.session.format_context()

            # Build the prompt
            if context:
                prompt = f"{context}\n\nUser: {content}"
            else:
                prompt = content

            # Add expand directive if requested
            if expand:
                prompt = make_expand_prompt(prompt)

            # Get system prompt
            all_models = [p.name for p in self.providers]
            system_prompt = get_system_prompt(provider.name, all_models)

            # Call the provider
            options = CompletionOptions(
                system_prompt=system_prompt,
                max_tokens=self.config.expand_max_tokens if expand else self.config.max_response_tokens,
            )

            response_text = await provider.generate(prompt, options)

            # Check for error
            if response_text.startswith("[Error]"):
                self.display.show_error(f"{provider.name}: {response_text}")
                return None

            return ChatMessage(
                role=MessageRole.ASSISTANT,
                content=response_text,
                model=provider.name,
                is_expanded=expand,
            )

        except Exception as e:
            self.display.show_error(f"{provider.name}: {e}")
            return None

    def _parse_mentions(self, content: str) -> tuple[list[str], str]:
        """
        Extract @mentions from message content.

        Returns:
            (mentioned_models, cleaned_content)
        """
        pattern = r"@(\w+)"
        mentions = re.findall(pattern, content)

        # Validate mentions against available models
        valid_mentions = []
        provider_names = [p.name.lower() for p in self.providers]

        for mention in mentions:
            if mention.lower() in provider_names:
                valid_mentions.append(mention.lower())

        # Don't remove mentions from content - keep them for context
        return valid_mentions, content

File: typescript/src/chat/room.ts
/**
 * Chat room orchestrator.
 */

import type { Provider, CompletionOptions } from '../providers/base.js';
import type { ChatConfig } from '../core/types.js';
import { DEFAULT_CHAT_CONFIG } from '../core/types.js';
import { ChatSession } from './session.js';
import { ChatDisplay } from './display.js';
import { CommandHandler } from './commands.js';
import { getSystemPrompt, makeExpandPrompt } from './prompts.js';

export class ChatRoom {
    public providers: Provider[];
    public config: ChatConfig;
    public session: ChatSession;
    private display: ChatDisplay;
    private commandHandler: CommandHandler;

    constructor(providers: Provider[], config?: Partial<ChatConfig>, session?: ChatSession) {
        this.providers = providers;
        this.config = { ...DEFAULT_CHAT_CONFIG, ...config };
        this.session = session || new ChatSession(providers.map(p => p.name));
        this.display = new ChatDisplay();
        this.commandHandler = new CommandHandler(this);
    }

    async start(): Promise<void> {
        this.display.showWelcome(this.providers.map(p => p.name));

        while (true) {
            try {
                const input = await this.display.getInput();
                const trimmed = input.trim();

                if (!trimmed) {
                    continue;
                }

                // Check for command
                const command = this.commandHandler.parse(trimmed);
                if (command) {
                    const result = await this.commandHandler.execute(command);
                    if (result.message) {
                        this.display.showCommandResult(result.message);
                    }
                    if (result.shouldExit) {
                        break;
                    }
                    continue;
                }

                // Parse @mentions
                const { content, targetModels } = this.parseMentions(trimmed);

                // Add user message to session
                this.session.addUserMessage(content);
                this.display.showUserMessage(content);

                // Send to models
                await this.sendMessage(content, targetModels);

            } catch (error) {
                if (error instanceof Error && error.message.includes('readline was closed')) {
                    break;
                }
                this.display.showError(error instanceof Error ? error.message : String(error));
            }
        }

        this.display.close();
    }

    async sendMessage(content: string, targetModelNames?: string[], expand: boolean = false): Promise<void> {
        // Determine which providers to use
        let targetProviders = this.providers;
        if (targetModelNames && targetModelNames.length > 0) {
            targetProviders = this.providers.filter(p =>
                targetModelNames.some(name =>
                    p.name.toLowerCase().includes(name.toLowerCase()) ||
                    name.toLowerCase().includes(p.name.toLowerCase())
                )
            );
            if (targetProviders.length === 0) {
                this.display.showError(`No matching models found for: ${targetModelNames.join(', ')}`);
                return;
            }
        }

        // Show thinking indicator
        this.display.showThinking(targetProviders.map(p => p.name));

        // Build context
        const context = this.session.formatContext(this.config.maxHistoryMessages);

        // Generate responses in parallel
        const responses = await Promise.all(
            targetProviders.map(async (provider) => {
                try {
                    const otherModels = this.providers
                        .filter(p => p.name !== provider.name)
                        .map(p => p.name);

                    let systemPrompt = getSystemPrompt(provider.name, otherModels);
                    if (expand) {
                        systemPrompt = makeExpandPrompt(systemPrompt);
                    }

                    const prompt = context ? `${context}\n\nUser: ${content}` : content;

                    const options: CompletionOptions = {
                        systemPrompt,
                        maxTokens: expand ? this.config.expandMaxTokens : this.config.maxResponseTokens,
                    };

                    const response = await provider.generate(prompt, options);
                    return { provider: provider.name, response, error: null };
                } catch (error) {
                    return {
                        provider: provider.name,
                        response: null,
                        error: error instanceof Error ? error.message : String(error)
                    };
                }
            })
        );

        this.display.clearThinking();

        // Display responses and add to session
        for (const result of responses) {
            if (result.error) {
                this.display.showError(`${result.provider}: ${result.error}`);
            } else if (result.response) {
                this.display.showModelResponse(result.provider, result.response);
                this.session.addAssistantMessage(result.response, result.provider, expand);
            }
        }
    }

    private parseMentions(input: string): { content: string; targetModels: string[] } {
        const mentionRegex = /@(\w+)/g;
        const mentions: string[] = [];
        let match;

        while ((match = mentionRegex.exec(input)) !== null) {
            mentions.push(match[1]);
        }

        // Remove @mentions from content
        const content = input.replace(/@\w+\s*/g, '').trim();

        return {
            content: content || input, // Fall back to original if nothing left
            targetModels: mentions,
        };
    }
}
